tljh/installer.py

```diff
@@ -154,58 +156,89 @@ def ensure_usergroups():
+    if not checksum:
+        raise ValueError(
+            f"Unsupported architecture: {arch}. TLJH only supports {','.join(MAMBAFORGE_CHECKSUMS.keys())}"
+        )
+    return installer_url, checksum
+
+
 def ensure_user_environment(user_requirements_txt_file):
     """
     Set up user conda environment with required packages
     """
     logger.info("Setting up user environment...")
+    # note: these must be in descending order
+    conda_upgrade_versions = {
+        # format: "conda version": (conda_version, mamba_version),
+        # mambaforge 4.10.3-7 (2023-03-21)
+        "22.11.1": (MAMBAFORGE_CONDA_VERSION, MAMBAFORGE_MAMBA_VERSION),
+        # tljh up to 0.2.0 (since 2021-10-18)
+        "4.10.3": ("4.10.3", "0.16.0"),
+    }
 
-    miniconda_old_version = "4.5.4"
-    miniconda_new_version = "4.7.10"
-    # Install mambaforge using an installer from
-    # https://github.com/conda-forge/miniforge/releases
-    mambaforge_new_version = "4.10.3-7"
-    # Check system architecture, set appropriate installer checksum
-    if os.uname().machine == "aarch64":
-        installer_sha256 = (
-            "ac95f137b287b3408e4f67f07a284357b1119ee157373b788b34e770ef2392b2"
-        )
-    elif os.uname().machine == "x86_64":
-        installer_sha256 = (
-            "fc872522ec427fcab10167a93e802efaf251024b58cc27b084b915a9a73c4474"
-        )
     # Check OS, set appropriate string for conda installer path
     if os.uname().sysname != "Linux":
         raise OSError("TLJH is only supported on Linux platforms.")
-    # Then run `mamba --version` to get the conda and mamba versions
-    # Keep these in sync with tests/test_conda.py::prefix
-    mambaforge_conda_new_version = "4.10.3"
-    mambaforge_mamba_version = "0.16.0"
-
-    if conda.check_miniconda_version(USER_ENV_PREFIX, mambaforge_conda_new_version):
-        conda_version = "4.10.3"
-    elif conda.check_miniconda_version(USER_ENV_PREFIX, miniconda_new_version):
-        conda_version = "4.8.1"
-    elif conda.check_miniconda_version(USER_ENV_PREFIX, miniconda_old_version):
-        conda_version = "4.5.8"
-    # If no prior miniconda installation is found, we can install a newer version
-    else:
+    found_conda = False
+    have_versions = conda.get_mamba_versions(USER_ENV_PREFIX)
+    have_conda_version = have_versions.get("conda")
+    if have_conda_version:
+        for check_version, conda_mamba_version in conda_upgrade_versions.items():
+            if V(have_conda_version) >= V(check_version):
+                found_conda = True
+                conda_version, mamba_version = conda_mamba_version
+                break
+
+    if not found_conda:
+        if os.path.exists(USER_ENV_PREFIX):
+            logger.warning(
+                f"Found prefix at {USER_ENV_PREFIX}, but too old or missing conda/mamba ({have_versions}). Rebuilding env from scratch!!"
+            )
+            # FIXME: should this fail? I'm not sure how destructive it is
```

Comment:
We should check this before merging! If in doubt I think failing and telling the admin to do X is OK, especially since this will be a breaking release, and we'll need to update jupyterhub-singleuser anyway. If it's easier I think we could say a new user env will always be created (and rename the old one)

Reply:
Test added. Re-running the installer does not wipe out the env, but it will upgrade many packages significantly. I'm sure there are cases where this can fail, but I think the behavior should be sensible and not too destructive.

I think it's an open question if/whether we should update conda/mamba from older versions, but at least this isn't changed by this PR, other than moving the 'way too old' bar up to 2021.

Reply:
If the situation is no worse than before then let's not worry about it in this PR

Reply:
This looks great to me now, marking this as resolved.